src/OmniSharp/Api/Navigation/OmnisharpController.NavigateUpDown.cs

```diff
@@ -0,0 +1,74 @@
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Mvc;
+using OmniSharp.Models;
+using System;
+
+namespace OmniSharp
+{
+    public partial class OmnisharpController
+    {
+        private Func<Request, FileMemberElement, GotoDefinitionResponse, bool> NavigateUpFunc = (request, node, qf) =>
+        {
+            bool retVal = false;
+            if (node.Location.Line < request.Line || (node.Location.Line == request.Line && node.Location.Column < request.Column))
+            {
+                qf.Line = node.Location.Line;
+                qf.Column = node.Location.Column;
+                retVal = true;
+            }
+            return retVal;
+        };
+
+        private Func<Request, FileMemberElement, GotoDefinitionResponse, bool> NavigateDownFunc = (request, node, qf) =>
+        {
+            bool retVal = false;
+            if (node.Location.Line > request.Line || (node.Location.Line == request.Line && node.Location.Column > request.Column))
+            {
+                qf.Line = node.Location.Line;
+                qf.Column = node.Location.Column;
+                retVal = true;
+            }
+            return retVal;
+        };
+
+        private async Task<IActionResult> Navigate(Request request, bool toBreak, Func<Request, FileMemberElement, GotoDefinitionResponse, bool> navigateFunc)
+        {
+            var stack = new List<FileMemberElement>(await StructureComputer.Compute(_workspace.GetDocuments(request.FileName)));
+            var ret = new GotoDefinitionResponse();
+            //Retain current line as result in case we dont need to navigate.
+            ret.Line = request.Line;
+            ret.Column = request.Column;
+            ret.FileName = request.FileName;
+
+            while (stack.Count > 0)
+            {
+                var node = stack[0];
+                stack.Remove(node);
+                //Need to call the passed in up or down delegate
+                var found = navigateFunc(request, node, ret);
+                if (toBreak && found)
+                {
+                    break;
+                }
+                stack.AddRange(node.ChildNodes);
+            }
+            return new ObjectResult(ret);
```

Comment:
Just return the object instead of the ObjectResult. It will make it simpler to test and we need POCOs for the stdout interface I believe.